### ISSUE TITLE
Can't create self-signed rsa-pss certificates with custom parameters using req command

### ISSUE THREAD

**Issue by tomato42** [openssl/openssl#2963]
Using current master (d1da335c5514f9b) I'm unable to parametrise the signature made by `openssl req -x509 -newkey` command, the signature always uses sha256 hash algorithm, mgf1 with sha256 and salt of maximum length:

```
openssl req -x509 -newkey rsa-pss -keyout rsassaPss-noRestrictions.key\
 -out rsassaPss-noRestrictions.crt -subj /CN=localhost -nodes -batch\
 -config ./openssl.cnf -pkeyopt rsa_keygen_bits:2048 -sigopt\
 rsa_padding_mode:pss -sigopt rsa_pss_saltlen:16 -sigopt rsa_mgf1_md:sha1 -sha1
```
(using config file with different `default_md` has no effect, the below example was created with one with sha1 as the default)
```
Generating a 2048 bit RSA-PSS private key
...............................................................+++
..............................+++
writing new private key to 'rsassaPss-noRestrictions.key'
-----
```
```
openssl x509 -in rsa-pss-no-restrictions.crt -noout -text
```
```
Certificate:
    Data:
        Version: 3 (0x2)
        Serial Number:
            bd:ec:ad:df:5e:82:78:80
    Signature Algorithm: rsassaPss         
         Hash Algorithm: sha256
         Mask Algorithm: mgf1 with sha256
          Salt Length: 0xDE
         Trailer Field: 0xBC (default)
        Issuer: CN = localhost
...
        Subject Public Key Info:
            Public Key Algorithm: rsassaPss
                RSA-PSS Public-Key: (2048 bit)
                Modulus:
...
                Exponent: 65537 (0x10001)
                No PSS parameter restrictions
        X509v3 extensions:
...
            X509v3 Basic Constraints: 
                CA:TRUE
    Signature Algorithm: rsassaPss         
         Hash Algorithm: sha256
         Mask Algorithm: mgf1 with sha256
          Salt Length: 0xDE
         Trailer Field: 0xBC (default)

 ...
```

**Comment by tomato42** [comment]
When the keygeneration is parametrised, the signature created is actually invalid:

```
openssl req -x509 -newkey rsa-pss -keyout ca.key -out ca.crt -subj /CN=CA\
 -nodes -batch -config /etc/pki/tls/openssl.cnf -pkeyopt rsa_keygen_bits:2048\
 -pkeyopt rsa_pss_keygen_md:sha256 -pkeyopt rsa_pss_keygen_saltlen:20\
 -sha256
```
```
openssl x509 -in ca.crt -noout -text
```
```
Certificate:
...
        Subject Public Key Info:
            Public Key Algorithm: rsassaPss
                RSA-PSS Public-Key: (2048 bit)
                Modulus:
...
                Exponent: 65537 (0x10001)
                PSS parameter restrictions:
                  Hash Algorithm: sha256
                  Mask Algorithm: mgf1 with sha256
                  Minimum Salt Length: 0x14 (default)
                  Trailer Field: 0xBC (default)
        X509v3 extensions:
 ...
            X509v3 Basic Constraints: 
                CA:TRUE
    Signature Algorithm: rsassaPss         
         Hash Algorithm: sha256
         Mask Algorithm: mgf1 with sha1 (default)
          Salt Length: 0x14 (default)
         Trailer Field: 0xBC (default)

...
```

Note that the mgf1 hash in signature is different than the one associated with the public key, fact pointed out by the verify command too:
```
openssl verify -CAfile ca.crt -check_ss_sig ca.crt
```
```
CN = CA
error 7 at 0 depth lookup: certificate signature failure
error ca.crt: verification failed
140629002618624:error:0408F098:rsa routines:pkey_rsa_ctrl:mgf1 digest not allowed:crypto/rsa/rsa_pmeth.c:514:
```

**Comment by bernd-edlinger** [comment]
I can not reproduce this with current trunk:

`./openssl req -x509 -newkey rsa-pss -nodes -config openssl.cnf -subj /CN=localhost -pkeyopt rsa_keygen_bits:2048 -sha1 | ./openssl x509 -noout -text`

```
=>     Signature Algorithm: rsassaPss         
         Hash Algorithm: sha1 (default)
         Mask Algorithm: mgf1 with sha1 (default)
          Salt Length: 0xEA
         Trailer Field: 0xBC (default)
```

**Comment by tomato42** [comment]
yes, the first issue is fixed, the second issue is still present (the one with key generation parametrisation)

also, can't override it with `-sigopt rsa_mgf1_md:sha256`

**Comment by snhenson** [comment]
I can confirm there's a bug with the second issue. Looking into it.

**Comment by snhenson** [comment]
Should be fixed by #3920. The bug was that the RSA_PSS_PARAMS structure contained a cache of the algorithm identifier to use with the mask. This was set whenever a key was decoded but not initially created. So if you created a key internally and used it immediately (as opposed to reading it from a file) it wouldn't use the correct parameters.

**Comment by mattcaswell** [comment]
According to the above comments this has been fixed. Closing.
